Add Boss Randomization

TriumphantBass opened the pull request:
**Introduction**
This PR adds a boss randomization setting that changes the placement of bosses.

Bosses XP, HP, and ATK will be scaled to the boss at the position they are replacing, unless the Scale Bosses setting is turned off.

This will not affect the item locations or when they are awarded (i.e. the vanilla boss rooms will have rewards when the bosses are defeated).

Changes *only active during boss rando*:
- Ravenlord arena has 6 moving platforms to allow full arena movement with only single jump (no dash, timestop, etc)
- Varndagroth item is rewarded *after* the fight

General changes:
- All bosses will record a `TSRando_<boss kill flag>`
This was originally just for boss rando because flags were getting set and unset frequently, but looking at completed save files, some of the bosses actually reset theirs naturally, so having a separate record helps keep room triggers more consistent.

**Known Issues**
- (Minor) - some bosses interrupt the replaced music with their own theme
- The Maw defeat cutscene was interfering with the Time Break effect, so if The Maw replaces Prince/Vol, it will sharp cut to ??? without performing a time break event.

TriumphantBass commented:
Okay, music fixed (save for Vol Terrilis) and setting added for disabling forced heals in boss rando, should be fully good to go

TriumphantBass commented on TsRandomizer/Randomisation/BestiaryManager.cs:
```diff
@@ -7,21 +11,506 @@
+		{
+			switch (bossId)
+			{
+				case 65:
```
Unfortunately the boss enum isn't in order, and lacks Vol Terrilis and Prince Nuvius
```
  internal enum EBossType
  {
    None,
    RoboKitty,
    Varndagroth,
    Bird,
    Demon,
    Maw,
    Sorceress,
    Shapeshift,
    Emperor,
    Sandman,
    Nightmare,
    Raven,
    Xarion,
    Zel,
    Cantoran,
  }
```

I can look around for something cleaner but there may not be many options bc I think the bestiary ids are just set in the bestiary data file

TriumphantBass commented on TsRandomizer/LevelObjects/Other/BossEnemy.cs:
```diff
@@ -0,0 +1,191 @@
+	[TimeSpinnerType("Timespinner.GameAbstractions.GameObjects.SandmanBoss")]
+	[TimeSpinnerType("Timespinner.GameObjects.Bosses.OtherBosses.NightmareBoss")]
+	// ReSharper disable once UnusedMember.Global
+	class BossEnemy: LevelObject
```
this line already has `: LevelObject`, were you saying to replace it with `class BossEnemy: Monster` or similar?

JarnoWesthof replied:
my sugggestion is to make it `: LevelObject<Monster>`

CaitSith2 commented:
Bugs:  Can't refight the bosses at sandman/nightmare a second time.  You just straight up enter sandmans and never teleport to whatever boss that replaced sandman.  The fix should be don't write the boss is dead flag if the boss replaced either sandman or nightmare.

Nice to have:  If a boss ends up at its vanilla location, unless its sandman as sandman with any other boss as nightmare, forego the teleportation stuff.

TriumphantBass commented:
While the vanilla boss at vanilla spot question is interesting, at the very least I'd like to leave it as-is so it's consistent.
While a boss randomized into the same room may look the same, the room refresh on both ends is needed for the flags and other cues to take full effect, and the "is fighting a randomized boss" controls things like deleting exits, supressing item drops, changing the peekability of Varndagroth, etc.
To limit risk of edge cases I would like to keep it as is, at least for the initial version release

Added endgame safety.

CaitSith2 commented:
Nightmare as Prince Nuvius or Nightmare as Vol Terrilis ends the game prematurely.  In the case of Archipelago, it prompts for !forfeit and !collect, but does NOT register to the archipelago server that the goal is complete.

TriumphantBass commented:
Added proper minion scaling

TriumphantBass commented:
Fixed AP goal state

CaitSith2 commented:
gas effect not present during the boss as The Maw fight, when missing a gas mask.  In other words, I didn't have to breath biscuits during the boss fight.

TriumphantBass commented:
Regression fixed 
![image](https://user-images.githubusercontent.com/32756996/171747549-fa0d3066-8dc0-4fce-8c8c-ac978ec30495.png)

CaitSith2 commented:
One thing I did note is that whatever level your oxygen was at, does NOT carry over to the boss room, but instead acts as if it is right full once again, and upon defeat of the boss, full Oxygen once again.

CaitSith2 commented:
Also, the boss room that replaces Feline Sentry SHOULD fill with gas, if The Maw is dead

CaitSith2 commented:
```
if (vanillaBossId == (int)EBossID.FelineSentry && level.GameSave.GetSaveBool("TSRando_IsBossDead_Maw"))
                FillRoomWithGas(level);
```

TriumphantBass commented:
I was following current rando (and vanilla) behavior where that room has an air pocket, but yes I can add that.

CaitSith2 commented:
And I found yet another bug.   I had Ifrit as Nightmare for my boss.  (This can also work for Ravenlord as Nightmare).  The key here, is that you clear out the two gyre boss rooms.

Now,  go through the Gyre Archive itself, then go to either the Ifrit spot or Ravenlord spot.  If you picked the one it randomly chose to initialize,  you will end up actually fighting either Ifrit or Ravenlord,  and if that is the as Nightmare fight, then you win upon defeating the boss.

TriumphantBass commented:
> https://github.com/JarnoWesthof/TsRandomizer/pull/81#issuecomment-1145521045

> And I found yet another bug. I had Ifrit as Nightmare for my boss. (This can also work for Ravenlord as Nightmare). The key here, is that you clear out the two gyre boss rooms.
> 
> Now, go through the Gyre Archive itself, then go to either the Ifrit spot or Ravenlord spot. If you picked the one it randomly chose to initialize, you will end up actually fighting either Ifrit or Ravenlord, and if that is the as Nightmare fight, then you win upon defeating the boss.

Do you have a seed number for this?

CaitSith2 commented:
[AllPlayers.zip](https://github.com/JarnoWesthof/TsRandomizer/files/8829314/AllPlayers.zip)  Save files containing the aformentioned situation. (since I ran this via Archipelago as part of my bug hunting.)

TriumphantBass commented:
Okay, reproduced and fixed. Entering a gyre portal wipes the IsBossDead for each which is part of why I added TSRando prefixes throughout.

Boss flags are now refreshed after each of the gyre portals
